src/NavierStokes.cpp

```diff
@@ -0,0 +1,758 @@
+  Vector<double> cell_rhs(dofs_per_cell);
+
+  std::vector<types::global_dof_index> dof_indices(dofs_per_cell);
+
+  system_matrix.copy_from(constant_matrix);
+  system_rhs = 0.0;
+
+  FEValuesExtractors::Vector velocity(0);
+  FEValuesExtractors::Scalar pressure(dim);
+
+  std::vector<Tensor<1, dim>> velocity_loc(n_q);
+  std::vector<Tensor<2,dim>> velocity_gradient_loc(n_q);
+  
+
+  unsigned int neumann_boundary_id;
+  if constexpr(dim==2) neumann_boundary_id=3;
+  if constexpr(dim==3) neumann_boundary_id=1;
+  
+	for (const auto &cell : dof_handler.active_cell_iterators())
+  {
+    if (!cell->is_locally_owned())
+      continue;
+
+    fe_values.reinit(cell);
+
+    non_linear_contribution = 0.0;
+    cell_rhs = 0.0;
+
+    fe_values[velocity].get_function_values(solution, velocity_loc);
+    fe_values[velocity].get_function_gradients(solution, velocity_gradient_loc);
+  
+  Tensor<1, dim> nonlinear_term;
+  Tensor<1, dim> transpose_term;
+    for (unsigned int q = 0; q < n_q; ++q)
+    {
+
+      Vector<double> forcing_term_loc(dim);
+      forcing_term.vector_value(fe_values.quadrature_point(q),
+                                forcing_term_loc);
+      Tensor<1, dim> forcing_term_tensor;
+      for (unsigned int d = 0; d < dim; ++d)
+        forcing_term_tensor[d] = forcing_term_loc[d];
+
+      for (unsigned int i = 0; i < dofs_per_cell; ++i)
+      {
+
+        if (use_skew){ // Skew symmetric representation of non linear term
+            for (unsigned int j = 0; j < dofs_per_cell; ++j)
+            { 
+              // Reset the nonlinear and transpose terms.
+              nonlinear_term = 0.0;
+              transpose_term = 0.0;
+              for (unsigned int k = 0; k < dim; k++) {
+                for (unsigned int l = 0; l < dim; l++) {
+                  // Calculate (u . nabla) u.
+                  nonlinear_term[k] += velocity_loc[q][l] *
+                                        fe_values[velocity].gradient(j, q)[k][l];
+                  // Calculate (nabla u)^T u.
+                  transpose_term[k] += velocity_loc[q][k] *
+                                        fe_values[velocity].gradient(j, q)[l][k];
+                }
+              }
+              // Add the skew-symmetric term (u . nabla) u + (nabla u)^T u to the matrix.
+              non_linear_contribution(i, j) +=
+                  0.5 * (scalar_product(nonlinear_term, fe_values[velocity].value(i, q)) +
+                          scalar_product(transpose_term, fe_values[velocity].value(i, q))) *
+                  fe_values.JxW(q);
+            }
+        }
+        else{
+          for (unsigned int j = 0; j < dofs_per_cell; ++j)
+                {
+                  // Semi-implicit treatment of the non linear term
+                  non_linear_contribution(i, j) += velocity_loc[q] * fe_values[velocity].gradient(j, q) *
+                                        fe_values[velocity].value(i, q) *
+                                        fe_values.JxW(q);
```

Comment:
I believe this term is not entirely correct. This corresponds to $(\mathbf{u}^n \cdot \nabla) \mathbf u \cdot \mathbf v$. In Einstein notation (repeated indices imply summation), this becomes $((\mathbf{u}^n \cdot \nabla)\mathbf u)_i \mathbf v_i = (\mathbf{u}^n \cdot \nabla) \mathbf u_i \mathbf v_i = \left( \mathbf{u}^n_j \frac{\partial}{\partial\mathbf{x}_j}\right) \mathbf{u}_i \mathbf{v}_i = \mathbf{u}^n_j \frac{\partial\mathbf u_i}{\mathbf x_j} \mathbf v_i = \mathbf u^n_j (\nabla\mathbf u)\_{ij} \mathbf v_i = (\nabla\mathbf u\mathbf u^n)_i \mathbf v_i = (\nabla\mathbf u \mathbf u^n)\cdot\mathbf v$.
Therefore, I believe this should be implemented as
```suggestion
non_linear_contribution(i, j) +=
  scalar_product(
    fe_values[velocity].gradient(j, q) * velocity_loc[q],
    fe_values[velocity].value(i, q)) * 
  fe_values.JxW(q);
```
This might explain the unusual behavior observed in your velocity field (see also my comments to the report).